src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] data;
+    private int size;
+
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkAndResize();
+        data[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkAndResize();
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't correctly add element by index "
+                    + index + "!");
+        }
+        System.arraycopy(data, index, data, index + 1, data.length - (index + 1));
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, "get");
+        return data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index, "replace");
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, "delete");
+        T temp = data[index];
+        if (index == 0 || size == data.length || index < size) {
+            System.arraycopy(data,index + 1, data, index, size - (index + 1));
+            size--;
+        }
+        return temp;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            size--;
+            return null;
+        }
+        for (int i = 0; i < size; i++) {
+            if (data[i] != null && data[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Item not found!");
```

Comment:
```suggestion
        for (int i = 0; i < size; i++) {
            if (data[i] == key || data[i] != null && data[i].equals(element)) {
                return remove(i);
            }
        }
        throw new NoSuchElementException("Item not found!");
```
your code works only when null is always in the last position